### ISSUE TITLE
Do object/array initializers execute `await`s sequentially or in parallel?

### ISSUE THREAD

**Issue by calebmer** [tc39/proposal-async-await#76]
Currently the only way I'm aware of executing promises in parallel is the `Promise.all()` function. However, it seems like an obvious optimization to add some intelligent detection to async/await functions which makes parallel `await`s possible. It also seems like object/array initializers are the perfect place to require this functionality/optimization.

While regular variable declarations are dependent on sequential order:

``` js
var a = 1
var b = a + 1
```

Object and array initializers are not:

``` js
var object = {
  a: 1,
  b: object.a + 1 // Throws a TypeError
}

var array = [1, array[0] + 1] // Throws a TypeError
```

So it would seem that the following parallel promise execution:

``` js
async function fetchData() {
  var data = await Promise.all([fetchDataA(), fetchDataB()])
  return {
    a: data[0],
    b: data[1]
  }
}
```

Could be sugared too the following without any performance penalties:

``` js
async function fetchData() {
  return {
    a: await fetchDataA(),
    b: await fetchDataB()
  }
}
```

The execution of `fetchDataB()` would not be blocked by the execution of `fetchDataA()` and when both are resolved the function resolves. The same could be said for arrays so that:

``` js
var data = await Promise.all([fetchDataA(), fetchDataB()])
```

Is synonymous in functionality to:

``` js
var data = [await fetchDataA(), await fetchDataB()]
```

---

So my question is, is this part of the spec? Do async/await functions currently operate like this? If so hooray for performance, if not is it a feasible addition? Is this an optimization only browsers would implement, or functionality that the spec mandates?

**Comment by getify** [comment]
> The execution of fetchDataB() would not be blocked by the execution of fetchDataA()

I believe it _would_ actually wait at the first `await` before moving to the `fetchDataB()` function call.

**Comment by calebmer** [comment]
@getify any reason it couldn't be parallel?

**Comment by getify** [comment]
> any reason it couldn't be parallel?

if the two functions happened to create side effects, and someone happened to be relying on those side-effects in a particular order.

**Comment by ljharb** [comment]
Just to clarify, `Promise.all` does not "execute" anything, nor does it involve any observable parallelism. It simply creates a promise that resolves after all of the provided promises have resolved, or, rejects after any of them has rejected.

Since `await` is syntax sugar for a nested `.then` call on a `Promise`, I'd expect that if you had two unrelated values, you'd explicitly use `Promise.all` and `await` the resulting promise, rather than needlessly setting up a dependency chain where there is no dependency.

**Comment by getify** [comment]
> `Promise.all` does not "execute" anything

Right, because the functions to produce its promises are already run before `Promise.all(..)` has a chance to run.

Whereas with `await`, which is a mechanism that's pausing local execution in-place... an expression like `[ await getPromise1(), await getPromise2() ]` is definitely going to evaluate `await getPromise1()` before it evaluates `await getPromise2()`.

Why? Because it's possible that the _creation_ of "promise 2" relies on the completion of "promise 1".

``` js
async function foobar() {
   var x = 0;

   function getPromise1() {
      return new Promise(function(resolve){
         setTimeout(function(){ x = 10; resolve(1); },100);
      });
   }
   function getPromise2() {
      return Promise.resolve(x);
   }

   return [ await getPromise1(), await getPromise2() ];
}
```

**Comment by qfox** [comment]
Brilliant example how NOT to code. Or how to shoot yourself in leg.

Btw:

``` js
async function foobar() {
  // ...
  return Promise.all([ getPromise1(), await getPromise2(), getPromise3() ]);
  // vs
  return [ await getPromise1(), await getPromise2(), await getPromise3() ];
}
```

Would be nice to have a sugar for `Promise.all`. But it shouldn't be a simple `await`. `await` is for `Promise.resolve` + fetching result from `.then`. But it can be changed later.

**Comment by getify** [comment]
> how NOT to code. Or how to shoot yourself in leg.

...or how to code in ways that cannot be automatically parallelized, which btw has dozens of years of precedence.

**Comment by qfox** [comment]
@getify Yeah, right. But still it's a bad example. If we wanna go functional programming way with monads, gonads, automatic parallelization or memoization we must use pure functions only. But you're right, we can't do that atm.

**Comment by maxnordlund** [comment]
Wasn't there a proposal about desugaring `await*` to `Promise.all`? As an analogue to `yield*`, then `[await foo(), await bar()]` would be sequential and `await* [foo(), bar()]` parallell.

**Comment by getify** [comment]
> Wasn't there a proposal about desugaring `await*`...

Yep, but it was abandoned. I personally disliked that idea because it made `await*` quite a bit different than `yield*`, which is a divergence of consistency that makes the explanation of `async..await` to devs tougher.

**Comment by ljharb** [comment]
Also, it made `Promise.all` special with no consideration for `Promise.race`, and other than the asterisk, `yield*` and `await*` weren't conceptually similar at all.
